include/rmm/device_buffer.hpp

```diff
@@ -0,0 +1,325 @@
+   * @brief Constructs a new `device_buffer` by deep copying the contents of
+   * another `device_buffer`.
+   *
+   * @note Only copies `other.size()` bytes from `other`, i.e., if `other.size()
+   * != other.capacity()`, then the size and capacity of the newly constructed
+   *`device_buffer` will be equal to `other.size()`.
+   *
+   * @throws std::bad_alloc If creating the new allocation fails
+   * @throws std::runtime_error if copying from `other` fails
+   *
+   * @param other The other `device_buffer` to deep copy
+   *---------------------------------------------------------------------------**/
+  device_buffer(device_buffer const& other)
+      : _size{other._size},
+        _capacity{other._size},
+        _stream{other._stream},
+        _mr{other._mr} {
+    _data = _mr->allocate(_size, _stream);
+    auto status =
+        cudaMemcpyAsync(_data, other._data, _size, cudaMemcpyDefault, _stream);
+
+    if (cudaSuccess != status) {
+      throw std::runtime_error{"Device memory copy failed."};
+    }
+  }
+
+  /**---------------------------------------------------------------------------*
+   * @brief Constructs a new `device_buffer` by moving the contents of another
+   * `device_buffer` into the newly constructed one.
+   *
+   * After the new `device_buffer` is constructed, `other` is modified to be
+   * empty, i.e., `data()` returns `nullptr`, and `size()` is zero.
+   *
+   * @throws Nothing
+   *
+   * @param other The `device_buffer` whose contents will be moved into the
+   * newly constructed one
+   *---------------------------------------------------------------------------**/
+  device_buffer(device_buffer&& other) noexcept
+      : _data{other._data},
+        _size{other._size},
+        _capacity{other._capacity},
+        _stream{other._stream},
+        _mr{other._mr} {
+    other._data = nullptr;
+    other._size = 0;
+    other._capacity = 0;
+    other._stream = 0;
+  }
+
+  /**---------------------------------------------------------------------------*
+   * @brief Copy assignment operator
+   *
+   * TODO: Decide if this should be deleted or not.
+   *
+   * @param other
+   * @return device_buffer&
+   *---------------------------------------------------------------------------**/
+  device_buffer& operator=(device_buffer const& other) = delete;
+  /*
+  {
+    if (&other != this) {
+      _mr->deallocate(_data, _size, _stream);
+      _size = other._size;
+      _stream = other._stream;
+      _mr = other._mr;
+      _data = _mr->allocate(_size, _stream);
+      auto status = cudaMemcpyAsync(_data, other._data, _size,
+                                    cudaMemcpyDefault, _stream);
+
+      if (cudaSuccess != status) {
+        throw std::runtime_error{"Device memory copy failed."};
+      }
+    }
+    return *this;
+  }
+  */
+
+  /**---------------------------------------------------------------------------*
+   * @brief Move assignment operator
+   *
+   * TODO: Decide if this should be deleted or not.
```

Comment:
Pros and cons?

Reply:
The decision to delete the move assignment operator will largely fall out from the decision on the copy assignment operator. I could see a case for implementing the move assignment operator, but leaving copy assignment deleted. Move assignment is relatively inexpensive, and it requires a user to explicitly opt-in by using std::move.

Reply:
I decided to keep them.